### PR TITLE
Add methods/tests

### PR DESCRIPTION
**Pull request by xiaoyun-matroid**
- Add new methods:
    - Now all methods on our current [public API documentation](https://www.matroid.com/docs/api/documentation) are covered

- Structure files/folders:
	- main file: `/matroid/client.py`
	- sub files: in folder `/matroid/src`
		- `accounts, collections, detectors, images, labels, streams, videos` contains API methods， categorized according to our documentation sections
		- `helpers` contains helper functions
	
	- tests: in folder `/test`
                 (`-s` is optional, for seeing printed info about the progress in terminal)
		- command for running all tests:
			```
			python -m pytest --client_id=yCzyYu2U1v3qPqAC --client_secret=NmBuz5lvnBW9Y9pbyGSAvfxEibMHxOOA --base_url=http://xiaoyun.devaz.matroid.com/api/v1 -s
			```
		- for running one test file:
			```
			python -m pytest test/test_streams.py --client_id=yCzyYu2U1v3qPqAC --client_secret=NmBuz5lvnBW9Y9pbyGSAvfxEibMHxOOA --base_url=http://xiaoyun.devaz.matroid.com/api/v1 -s
			```
There are some changes to our route files in my another PR in the main repo, so some tests will only pass on my dev website on that branch

### CODE REVIEW

File: matroid/src/collections.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        data = {
+            'includeCollectionInfo': 'true' if include_collection_info else ''
+        }
+        return requests.request(method, endpoint, **{'headers': headers, 'data': data})
+    except Exception as e:
+        raise error.APIConnectionError(message=e)
+
+
+@api_call(error.InvalidQueryError)
+def query_by_detection_scores(self, task_id, thresholds, num_results):
```
Comment (john-matroid) [inline matroid/src/collections.py]: Let's add some top level comments for these functions as well. If it isn't clear, `query_by_detection_scores` takes in a map of thresholds, and returns media in the collection with detections above those thresholds. `query_by_image` takes in an image file or url and returns similar media from the collection..

**Review by john-matroid — COMMENTED** [review]
Woohoo, awesome new functionality and way better file structure!

File: matroid/error.py [hunk 2]
```diff
@@ -1,60 +1,61 @@
...
+        else:
+            try:
+                super(APIError, self).__init__(response.text)
+            except AttributeError:
+                super(APIError, self).__init__(
+                    'No response or error message provided')
 
-    if response:
-      self.response = response
+        if response:
```
Comment (edfanchermat) [inline matroid/error.py]: The standard at Matroid seems to be 2 spaces vs. the default 4.

File: matroid/src/collections.py (hunk 1 above)
Comment (edfanchermat) [inline matroid/src/collections.py]: Also, it might be nice to add a link to the corresponding rest documentation. 
https://staging.dev.matroid.com/docs/api/index.html#api-Collections-PostApiVersionCollectionTasksTaskidScoresQuery in this case.

File: matroid/src/helpers.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        elif UserErr:
+            raise UserErr(response)
+        else:
+            raise error.APIError(response)
+    elif code == 'media_err':
+        raise error.MediaError(response)
+    elif status / 100 == 5 and code == 'server_err':
+        raise error.ServerError(response)
+    elif status / 100 != 2:
+        raise error.APIError(response)
```
Comment (edfanchermat) [inline matroid/src/helpers.py]: Can there be any other error codes? Might it end up just dropping through and failing silently?

File: matroid/src/labels.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    """Get annotations. Requires processing=false. Note: you need to provide at least one of the three ids to query"""
+    (endpoint, method) = self.endpoints['get_annotations']
+
+    detector_id = options.get('detector_id')
+    label_ids = options.get('label_ids')
+    image_id = options.get('image_id')
+
+    if not detector_id and not label_ids and not image_id:
+        raise error.InvalidQueryError(
+            message='Missing required parameter: detector_id or label_ids o rimage_id')
```
Comment (edfanchermat) [inline matroid/src/labels.py]: Typo: o rimage_id

File: matroid/src/detectors.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        bbox.csv
+    """
+    MAX_LOCAL_ZIP_SIZE = 300 * 1024 * 1024
+
+    (endpoint, method) = self.endpoints['create_detector']
+
+    try:
+        headers = {'Authorization': self.token.authorization_header()}
+        data = {'name': name, 'detector_type': detector_type}
+        with self.filereader.get_file(zip_file) as file_to_upload:
```
Comment (edfanchermat) [inline matroid/src/detectors.py]: I don't see where this file gets closed. There are a couple of others like this.

File: test/test_accounts.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import unittest
```
Comment (edfanchermat) [inline test/test_accounts.py]: Any chance we could use pytest instead? Then it would be similar to the selenium tests.

File: test/test_collections.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import time
+from datetime import datetime
+
+from .tests_helpers import set_up_client, print_test_title, print_case_pass, EVERYDAY_OBJECT_DETECTOR_ID, TEST_IMAGE_URL
+
+COLLECTION_NAME = 'py-test-collection-{}'.format(datetime.now())
+S3_BUCKET_URL = 's3://bucket/m-test-public/'
+
+
+class TestCollections(unittest.TestCase):
```
Comment (edfanchermat) [inline test/test_collections.py]: It might be good to have some negative tests:
  bad collection id
  bad label
  bad task id
  bad bounding box

File: test/test_detectors_labels.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+import time
+import unittest
+import os
+
+from .tests_helpers import set_up_client, print_test_title, print_case_pass, TEST_IMAGE_FILE, EVERYDAY_OBJECT_DETECTOR_ID
+
+DETECTOR_TEST_ZIP = os.getcwd() + '/test/test_file/cat-dog-lacroix.zip'
+
+
+class TestDetectorsAndLabels(unittest.TestCase):
```
Comment (edfanchermat) [inline test/test_detectors_labels.py]: Here too, it would be nice to have some simple tests for bad id's, basically any parameter that could be bad. Really goes for all of the test modules...

File: test/test_streams.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
+from datetime import datetime
+import time
+import os
+import unittest
+
+from .tests_helpers import set_up_client, print_test_title, print_case_pass, EVERYDAY_OBJECT_DETECTOR_ID, TEST_VIDEO_URL
+
+
+class TestStreams(unittest.TestCase):
```
Comment (edfanchermat) [inline test/test_streams.py]: Streaming has special permissions, so you might try a test where the user doesn't have streaming detect privileges.

File: test/test_vidoes.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import unittest
```
Comment (edfanchermat) [inline test/test_vidoes.py]: typo in the filename: test_videos.

**Review by edfanchermat — COMMENTED** [review]
Nice. Lots of tests. :) 
Mostly small things and one big request (pytest)

Thanks,

Ed.

File: test/test_collections.py (hunk 7 above)
Comment (xiaoyun-matroid) [reply test/test_collections.py]: At first, I thought the tests here are just for checking if we are hitting the correct API routes, and we can get some response from the server, but what kind of response we get is covered by the public API tests in our main repo, but I can add some negative test cases here too!

File: test/test_accounts.py (hunk 6 above)
Comment (xiaoyun-matroid) [reply test/test_accounts.py]: Sure! I'll take a look at pytest and try converting the tests. Thanks for all the comments! I'll try changing a small portion into pytest and make sure they work, then fix the problems/typos according to the comments!

File: matroid/src/detectors.py (hunk 5 above)
Comment (xiaoyun-matroid) [reply matroid/src/detectors.py]: The is what we already have in our repo, I copied the code to the new file so I didn't check closely.. Thanks for reminding me to check and make sure all file gets closed! I'll check all the places I used files and fix them after converting the tests

File: matroid/src/images.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    except IOError as e:
+        raise e
+    except error.InvalidQueryError as e:
+        raise e
+    except Exception as e:
+        raise error.APIConnectionError(message=e)
+
+
+@api_call(error.InvalidQueryError)
+def localize_image(self, localizer, localizer_label, **options):
```
Comment (grant-matroid) [inline matroid/src/images.py]: Nice - I think this was missing in the previous version, even though `localize_image` was in the docs for the API.

File: matroid/src/detectors.py (hunk 5 above)
Comment (dannyjeck-matroid) [reply matroid/src/detectors.py]: With statements should close the file automatically once the context gets exited.

Comment (dannyjeck-matroid) [reply matroid/src/detectors.py]: https://stackoverflow.com/questions/9885217/in-python-if-i-return-inside-a-with-block-will-the-file-still-close

**Review by grant-matroid — APPROVED** [review]
Covered my issues with the API in
- classifying more than 10 images ( not allowed, but also this restriction was not in docs)
- Functions that were not implemented but in the docs

Thanks Xiaoyun!

File: matroid/src/collections.py (hunk 1 above)
Comment (xiaoyun-matroid) [reply matroid/src/collections.py]: Added!

File: matroid/error.py (hunk 2 above)
Comment (xiaoyun-matroid) [reply matroid/error.py]: Fixed!

File: matroid/src/collections.py (hunk 1 above)
Comment (xiaoyun-matroid) [reply matroid/src/collections.py]: Added URLs as comments above each method!

File: matroid/src/helpers.py (hunk 3 above)
Comment (xiaoyun-matroid) [reply matroid/src/helpers.py]: This covers 2xx, 4xx, and 5xx, which seems to be enough for most cases, maybe I can add an `OtherError` or `UnknownError` to catch all other errors, and if we get that type of errors we can add specific ones for them later?

File: matroid/src/labels.py (hunk 4 above)
Comment (xiaoyun-matroid) [reply matroid/src/labels.py]: fixed!

File: test/test_accounts.py (hunk 6 above)
Comment (xiaoyun-matroid) [reply test/test_accounts.py]: I changed all tests into pytest, this is the first time I've ever used pytest, please let me know if there's anything I could improve! Thanks!

File: test/test_vidoes.py (hunk 10 above)
Comment (xiaoyun-matroid) [reply test/test_vidoes.py]: Fixed!

File: test/test_streams.py (hunk 9 above)
Comment (xiaoyun-matroid) [reply test/test_streams.py]: Do you mean users who do not have enough credits to monitor a stream?

File: test/test_detectors_labels.py (hunk 8 above)
Comment (xiaoyun-matroid) [reply test/test_detectors_labels.py]: I started by adding some negative tests to some POST methods, should I aim for adding one negative test to each method?

File: test/test_streams.py (hunk 9 above)
Comment (edfanchermat) [reply test/test_streams.py]: On the admin page for a user, there's a permission settings for can detect a stream. 
image
It's probably not a big deal, I was just thinking the error message on the api might not be great.

File: matroid/src/helpers.py (hunk 3 above)
Comment (dannyjeck-matroid) [reply matroid/src/helpers.py]: Do we get other errors if the user tries to use http instead of https?

Comment (xiaoyun-matroid) [reply matroid/src/helpers.py]: I tried sending requests of `get account info/create stream/create monitoring/get monitoring result/kill monitoring/delete monitoring` to url `http://www.matroid.com/api/v1` and `http://staging.dev.matroid.com/api/v1`, they seem to work just the same🤔

Comment (dannyjeck-matroid) [reply matroid/src/helpers.py]: probably getting redirected then. That's fine.

**Comment by john-matroid** [comment]
Info on actually updating the client:

https://docs.google.com/document/d/1BR_nM8xD_gLJqFZIxmTJCD-lRBU_iD15pig8kg4zBQc/edit?usp=sharing

**Review by john-matroid — APPROVED** [review]

**Comment by john-matroid** [comment]
@grant-matroid & @xiaoyun-matroid you guys should sync on what else we need to add & when / how to push out an update.

File: matroid/src/streams.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
+import requests
+import json
+
+from matroid import error
+from matroid.src.helpers import api_call
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Streams-PostStreams
+@api_call(error.InvalidQueryError)
+def create_stream(self, stream_url, stream_name):
```
Comment (grant-matroid) [inline matroid/src/streams.py]: Xiaoyun, I mentioned this in slack, but I think we should try to be consistent with the names of the arguments to the routes across API clients. 

For example, in the docs here https://www.matroid.com/docs/api/index.html#api-Streams-PostStreams, the arg names are `name` and `url` where here they are `stream_url` and `stream_name`.

I know this is a small detail, but I think keeping things consistent will make the docs easier to read and use.

File: matroid/src/detectors.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    return requests.request(method, endpoint, **{'headers': headers})
+  except Exception as e:
+    raise error.APIConnectionError(message=e)
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Detectors-Search
+@api_call(error.InvalidQueryError)
+def search_detectors(self, **query):
```
Comment (grant-matroid) [inline matroid/src/detectors.py:186]: Are we going to use `search_detectors` instead of `list_detectors` for other clients?

File: matroid/src/labels.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
+import requests
+import json
+
+from matroid import error
+from matroid.src.helpers import api_call, batch_file_request
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Labels-PostDetectorsDetector_idLabels
+@api_call(error.InvalidQueryError)
+def create_label_with_images(self, detector_id, name, image_files, **options):
```
Comment (grant-matroid) [inline matroid/src/labels.py]: I think the docs describe a "destination" arg where you can specify "images", deselectedImages", "negativeImages", etc... do we want to support that here?

Docs: https://www.matroid.com/docs/api/index.html#api-Labels-PostDetectorsDetector_idLabels

File: matroid/src/labels.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+def update_annotations(self, detector_id, label_id, images, **options):
+  (endpoint, method) = self.endpoints['update_annotations']
+  endpoint = endpoint.replace(':detector_id', detector_id)
+  endpoint = endpoint.replace(':label_id', label_id)
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    data = {
+        'images': json.dumps(images),
+        'destination': options.get('destination')
```
Comment (grant-matroid) [inline matroid/src/labels.py]: In the docs here https://www.matroid.com/docs/api/index.html#api-Labels-UpdateAnnotations `images` does not seem like it's optional, but `destination` is optional. 

Maybe we should separate them here to reflect that?

File: matroid/src/labels.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+def update_label_with_images(self, detector_id, label_id, image_files, **options):
+  """Requires processing=false. Updates label asynchronously (turn processing to true)"""
+  (endpoint, method) = self.endpoints['update_label_with_images']
+  endpoint = endpoint.replace(':detector_id', detector_id)
+  endpoint = endpoint.replace(':label_id', label_id)
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    data = {
+        'destination': options.get('destination'),
```
Comment (grant-matroid) [inline matroid/src/labels.py]: Similar here : `destination` and `bboxes` are optional in the docs. Could use `update`

File: matroid/src/streams.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
+import requests
+import json
+
+from matroid import error
+from matroid.src.helpers import api_call
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Streams-PostStreams
+@api_call(error.InvalidQueryError)
+def create_stream(self, url, name):
```
Comment (grant-matroid) [inline matroid/src/streams.py]: Do we want to support the optional args here? 
https://www.matroid.com/docs/api/index.html#api-Streams-PostStreams

File: matroid/src/streams.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+  if options.get('format') == 'csv' and self.json_format:
+    print('cannot return csv format when json_format True is specified upon API object initialization')
+    print ('requesting JSON format...')
+    options['format'] = 'json'
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    params = {
+        'format': options.get('format'),
+        'status_only': 'true' if options.get('status_only') else 'false'
```
Comment (grant-matroid) [inline matroid/src/streams.py]: Again, could use `update`

File: matroid/src/streams.py [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+@api_call(error.InvalidQueryError)
+def monitor_stream(self, stream_id, detector_id, **options):
+  (endpoint, method) = self.endpoints['monitor_stream']
+  endpoint = endpoint.replace(':detector_id', detector_id)
+  endpoint = endpoint.replace(':stream_id', stream_id)
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    data = {
+        'thresholds': json.dumps(options.get('thresholds')),
```
Comment (grant-matroid) [inline matroid/src/streams.py]: Thresholds is not marked as optional in the docs 
https://www.matroid.com/docs/api/index.html#api-Streams-PostStreamsStream_idMonitorDetector_id

File: matroid/src/streams.py [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+        'startTime': options.get('start_time'),
+        'endTime': options.get('end_time'),
+        'endpoint': options.get('endpoint'),
+        'taskName': options.get('task_name')
```
Comment (grant-matroid) [inline matroid/src/streams.py]: Another place to group optional args via `update`

File: matroid/src/videos.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
+import os
+import requests
+
+from matroid import error
+from matroid.src.helpers import api_call
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Videos-PostDetectorsDetector_idClassify_video
+@api_call(error.InvalidQueryError)
+def classify_video(self, detector_id, url=None, file=None):
```
Comment (grant-matroid) [inline matroid/src/videos.py]: Looks like we aren't checking for optional args `fps` and `annotationThresholds`?

File: matroid/src/videos.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    print('requesting JSON format...')
+    format = 'json'
+
+  endpoint = endpoint.replace(':key', video_id)
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    params = {
+        'videoId': video_id,
+        'threshold': threshold,
```
Comment (grant-matroid) [inline matroid/src/videos.py]: Could group the optional args `threshold` and `annotations` via `update`

File: matroid/src/images.py [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+  if not url and not file:
+    raise error.InvalidQueryError(
+        message='Missing required parameter: file or url')
+
+  endpoint = endpoint.replace(':key', detector_id)
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    data = {'detector_id': detector_id}
+    data.update(options)
```
Comment (grant-matroid) [inline matroid/src/images.py:29]: Seems like the safest way to handle optional args, and keep them in the same try block!

File: matroid/src/streams.py [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    params = {
+        'stream_id': options.get('stream_id'),
+        'monitoring_id': options.get('monitoring_id'),
+        'detector_id': options.get('detector_id'),
+        'name': options.get('name'),
+        'start_time': options.get('start_time'),
+        'end_time': options.get('end_time'),
+        'state': options.get('state'),
```
Comment (grant-matroid) [inline matroid/src/streams.py]: Another place to group optional args via `update`

File: matroid/src/detectors.py [hunk 25]
```diff
@@ -0,0 +1,198 @@
+import os
+import requests
+
+from matroid import error
+from matroid.src.helpers import api_call
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Detectors-PostDetectors
+@api_call(error.InvalidQueryError)
+def create_detector(self, file, name, detector_type):
```
Comment (grant-matroid) [inline matroid/src/detectors.py]: I think we still need to update this to support updating a pending detector, but that route might need to changed in matroid repo, or we can remove it from the docs. Should be a separate PR

File: matroid/src/collections.py [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+        'detectorId': detector_id,
+        'fileTypes': file_types
+    }
+    return requests.request(method, endpoint, **{'headers': headers, 'data': data})
+  except Exception as e:
+    raise error.APIConnectionError(message=e)
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Collections-PostCollections
+@api_call(error.InvalidQueryError)
+def create_collection(self, name, url, source_type):
```
Comment (grant-matroid) [inline matroid/src/collections.py]: `source_type` maybe should be `sourceType` to match route arg and docs?

File: matroid/src/collections.py [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+
+  try:
+    headers = {'Authorization': self.token.authorization_header()}
+    return requests.request(method, endpoint, **{'headers': headers})
+  except Exception as e:
+    raise error.APIConnectionError(message=e)
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Collections-PostCollectionTasksTaskidKill
+@api_call(error.InvalidQueryError)
+def kill_collection_index(self, task_id, include_collection_info):
```
Comment (grant-matroid) [inline matroid/src/collections.py]: `include_collection_info` maybe should match `includeCollectionInfo`?

File: matroid/src/collections.py [hunk 28]
```diff
@@ -0,0 +1,184 @@
...
+        'includeCollectionInfo': 'true' if include_collection_info else ''
+    }
+    return requests.request(method, endpoint, **{'headers': headers, 'data': data})
+  except Exception as e:
+    raise error.APIConnectionError(message=e)
+
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Collections-PostApiVersionCollectionTasksTaskidScoresQuery
+@api_call(error.InvalidQueryError)
+def query_collection_by_scores(self, task_id, thresholds, num_results):
```
Comment (grant-matroid) [inline matroid/src/collections.py]: `num_results` should match `numResults`

File: matroid/src/collections.py [hunk 29]
```diff
@@ -0,0 +1,184 @@
...
+        'thresholds': json.dumps(thresholds),
+        'numResults': num_results
+    }
+    return requests.request(method, endpoint, **{'headers': headers, 'data': data})
+  except Exception as e:
+    raise error.APIConnectionError(message=e)
+
+# https://staging.dev.matroid.com/docs/api/index.html#api-Collections-PostApiCollectionTasksTaskidImageQuery
+@api_call(error.InvalidQueryError)
+def query_collection_by_image(self, task_id, bounding_box=None, url=None, file=None, **options):
```
Comment (grant-matroid) [inline matroid/src/collections.py]: `bounding_box` should match `boundingBox`

**Review by grant-matroid — COMMENTED** [review]
Mostly checked for consistencies between the python args and the route args outlined in the docs and outlined in this PR https://github.com/matroid/matroid/pull/7935

Thanks Xiaoyun!

File: matroid/src/detectors.py (hunk 13 above)
Comment (xiaoyun-matroid) [reply matroid/src/detectors.py:186]: Yes, I feel this name is better than "list_detectors", and we already have "search_streams" and "search_monitorings" 😅
